shared/android/app/src/main/java/io/keybase/ossifrage/MainActivity.java

```diff
@@ -37,142 +49,282 @@
+    // Here we are just reading from the notification bundle.
+    // If other sources start the app, we can get their intent data the same way.
+    Bundle bundleFromNotification = intent.getBundleExtra("notification");
+
+    // TODO this doesn't work and didn't work before
+    String fromShareText = intent.getStringExtra(Intent.EXTRA_TEXT);
+    if (fromShareText == null) {
+      fromShareText = "";
     }
+    String finalFromShareText = fromShareText;
+
+    Uri uri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
 
-    @Override
-    protected void onDestroy() {
-        super.onDestroy();
-        Keybase.appWillExit(new KBPushNotifier(this));
+    // If there isn't any data we care about, let's just return
+    if (bundleFromNotification == null && fromShareText.isEmpty() && uri == null) {
+      return;
     }
 
-    @Override
-    public void onNewIntent(Intent intent) {
-        NativeLogger.info("new Intent: " + intent.getAction());
-        super.onNewIntent(intent);
-        setIntent(intent);
+    // Closure like class so we can keep our emit logic together
+    class Emit implements Runnable {
+      private DeviceEventManagerModule.RCTDeviceEventEmitter emitter;
+
+      Emit(DeviceEventManagerModule.RCTDeviceEventEmitter emitter) {
+        this.emitter = emitter;
+      }
+
+      @Override
+      public void run() {
+        assert emitter != null;
+        // If there are any other bundle sources we care about, emit them here
+        if (bundleFromNotification != null) {
+          emitter.emit("initialIntentFromNotification", Arguments.fromBundle(bundleFromNotification));
+        }
+
+        if (!finalFromShareText.isEmpty()) {
+          WritableMap args = Arguments.createMap();
+          args.putString("text", finalFromShareText);
+          emitter.emit("onShareText", args);
+        }
+
+        if (uri != null) {
+          String filePath = readFileFromUri(getReactContext(), uri);
+          if (filePath != null) {
+            WritableMap args = Arguments.createMap();
+            args.putString("localPath", filePath);
+            emitter.emit("onShareData", args);
+          }
+        }
+      }
     }
 
-    /**
-     * Returns the name of the main component registered from JavaScript. This is
-     * used to schedule rendering of the component.
-     */
-    @Override
-    protected String getMainComponentName() {
-        return "Keybase";
+    // We need to run this on the main thread, as the React code assumes that is true.
+    // Namely, DevServerHelper constructs a Handler() without a Looper, which triggers:
+    // "Can't create handler inside thread that has not called Looper.prepare()"
+    Handler handler = new Handler(Looper.getMainLooper());
+    handler.post(() -> {
+      // Construct and load our normal React JS code bundle
+      ReactInstanceManager mReactInstanceManager = ((ReactApplication) getApplication()).getReactNativeHost().getReactInstanceManager();
```

Comment:
Why is this called `mReactInstanceManager`? It's not a member variable, it's a local variable...

Reply:
copy/pasted, but good point :+1: